### DISCUSSION TITLE
trapping effect in relative permeability calculation of two phase?

### DISCUSSION THREAD

**Discussion by Gal-Tzvik** [PMEAL/OpenPNM#2916] [Q&A]
Following the proposed algorithm given in the examples, I witnessed that this does not take in to account the effect of the trapping within the invasion. After trying to add the trapping effect there seemed to be some discrepancies between the drainage results and the relative permeability limits of calculation. I was wondering if maybe there is something wrong with my code/am i not looking at this properly considering he evaluation of two different properties, or is this due to the inherent properties of the library functions.

As an example, you can see my drainage curve with and without trapping, where the non wetting phases invades only ~80% percent of the total volume when considering trapping but when calculating the relative permeability the invasion sequence only reaches ~55%.

Drainage Curve
![image](https://github.com/PMEAL/OpenPNM/assets/165297861/dfeca7c9-c533-4873-a646-6e77e0dc0fcb)

No trapping - RP
![image](https://github.com/PMEAL/OpenPNM/assets/165297861/b763e69c-4670-4484-9c73-121e5779179b)

RP with trapping 
![image](https://github.com/PMEAL/OpenPNM/assets/165297861/8f655afd-fc07-452d-bbe0-4ea4c35eb06b)

**Comment by jgostick** [comment]
These should both be using the same invasion sequence numbers, so not sure what the discrepancy is, but my first guess would be that the x-axis in the rel perm curve is not including the volume of throats, only of the pores.

**Reply by Gal-Tzvik** [reply]
Both simulations are done on the exact same network with the same inlet and outlet boundaries. Moreover, the rel perm specifically calculates the throat and pore volume as presented in the example for the occupancy update. Therefore i dont understand what can cause this difference. The only thing i can think of is the stokes flow solver.

**Reply by jgostick** [reply]
Me neither, that's why I was asking for more information.  Perhaps the pc curve is not using the throat volume then?

**Reply by Gal-Tzvik** [reply]
Ok, due to what you said, i evaluated the invasion sequence, and there is a difference between them. The drainage algorithm gives a different invasion sequence from the ip sequence, For some reason the inlets in the drainage algorithm arent considered the first pores to be invaded. I will go over the source code and update. Thanks for the quick response.

**Reply by jgostick** [reply]
Yes, there will definitely be a difference between them.  Invasion goes one pore at a a time, so it able to also capture trapping one pore at a time (ie. more correctly), while drainage goes one pressure step at a time, so trapping is only evaluated after each step hence missing many of the trapping events.  Note that drainage is a subset of invasion, or more generally the drainage curve represents the envelop of peak pressures of the invasion curve.

**Reply by Gal-Tzvik** [reply]
I am aware of this, therefore i dont understand why i have differences in a general characteristic such as the maximum non wetting phase sat.

**Reply by jgostick** [reply]
As you mentioned, the inlet's are treated differently in each case, which may lead to subtle changes in the pattern, but I would not expect this to be huge (assuming the network is large). Regarding my other point about the differences in trapping between invasion vs drainage, [this paper](https://doi.org/10.1016/j.cageo.2021.104978) has some more discussion in terms of image-based invasion. The point is that drainage will always under-estimate trapping since it only evaluates it after many pores are filled, whereas invasion evaluates is after each pore, so catches every trapping event.

**Reply by Gal-Tzvik** [reply]
Thank you i will definitely look into it. But just so i understand what you mean, the findings considering trapping with the ip algorithm should present more accurate results?

**Reply by jgostick** [reply]
The IP is more rigorous when looking at trapping, but this is only from a numerical standpoint.  In physical terms trapping is a bit more complicated since the trapped phase can evaporate or move around via corner flow.  So while IP does a better job at finding trapped defending phase, I can't say it's more or less accurate compared to reality.

**Reply by Gal-Tzvik** [reply]
Thank you very much for the support!
